### PR TITLE
Port samplers to core

### PR DESCRIPTION
**Pull request by mattdangerw**
🚧 This is an experimental feature branch, more details soon. 🚧

This one is fairly rote, tf -> ops. No beam or contrastive yet, hoping to do that soon.

### CODE REVIEW

File: keras_nlp/backend/ops.py [hunk 1]
```diff
@@ -42,7 +42,8 @@ def take_along_axis(x, indices, axis=None):
             if static_shape[i] != 1:
                 squeezable = False
         if squeezable:
-            indices = tf.squeeze(indices, leftover_axes)
+            if leftover_axes:
+                indices = tf.squeeze(indices, leftover_axes)
             return tf.gather(x, indices, batch_dims=axis)
```
Comment (chenmoneygithub) [inline keras_nlp/backend/ops.py:47]: wondering - why is the `gather` call in the `if squeezable` branch?

File: keras_nlp/samplers/beam_sampler.py [hunk 2]
```diff
@@ -101,6 +102,8 @@ def __init__(
         return_all_beams=False,
         **kwargs,
     ):
+        assert_tf_backend(self.__class__.__name__)
```
Comment (chenmoneygithub) [inline keras_nlp/samplers/beam_sampler.py:105]: maybe add a comment "Beam search only works in Tensorflow backend + short explanation"?

File: keras_nlp/samplers/beam_sampler_test.py [hunk 3]
```diff
@@ -31,36 +34,37 @@ def setUp(self):
         self.vocab_size = len(self.int_lookup)
 
         def next(prompt, cache, index):
-            batch_size = tf.shape(prompt)[0]
+            batch_size = ops.shape(prompt)[0]
```
Comment (chenmoneygithub) [inline keras_nlp/samplers/beam_sampler_test.py:37]: Not related to this PR, now I am starting to feel that it would be harder for contributors to write PRs because our operations comes from `ops` instead of the usual libraries like `tf` or `torch`. 
If things don't work out at ops level, contributors will need to write the ops correctly.

File: keras_nlp/samplers/contrastive_sampler.py [hunk 4]
```diff
@@ -78,6 +79,8 @@ def __init__(
         seed=None,
         **kwargs,
     ):
+        assert_tf_backend(self.__class__.__name__)
```
Comment (chenmoneygithub) [inline keras_nlp/samplers/contrastive_sampler.py:82]: Same here.

**Review by chenmoneygithub — APPROVED** [review]

File: keras_nlp/samplers/beam_sampler_test.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/samplers/beam_sampler_test.py:37]: Agreed! Something to work out over the long term.

File: keras_nlp/backend/ops.py (hunk 1 above)
Comment (jbischof) [reply keras_nlp/backend/ops.py:47]: This line and its scope is not modified by the PR 🤔 

Might want a separate PR to clean up if it's an issue.

File: keras_nlp/samplers/beam_sampler_test.py (hunk 3 above)
Comment (jbischof) [reply keras_nlp/samplers/beam_sampler_test.py:37]: This is an important point: Keras Core development is more expensive than tf.keras! Everything is a tradeoff for sure

File: keras_nlp/backend/ops.py (hunk 1 above)
Comment (mattdangerw) [reply keras_nlp/backend/ops.py:47]: Hopefully this is a short lived hack. We would like a `take_along_axis` in keras-core that just works with dynamic shape.

For now the hack is, if indices is all squeezable after `axis`, we can use `gather` directly. Otherwise we would need to write a complex `gather_nd` thing, which would effectively be rewriting `tfnp.take_along_axis`. If we do that, we should just land in keras-core.

Thankfully, this is the last of these workaround we need in `keras-nlp` for the `ops` layer, which is pretty cool!

**Review by jbischof — APPROVED** [review]
Thanks!

File: keras_nlp/backend/ops.py (hunk 1 above)
Comment (mattdangerw) [reply keras_nlp/backend/ops.py:47]: What's being fixed here is a bug in the workaround. I had thought `tf.squeeze(x, axis=())` would be a no-op, but instead it means to squeeze all squeezable axes.

This would break some squeeze calls in the `samplers`, and as a result some of those tests.

My hope is to just spend time patching keras-core itself, rather than the workaround here too much.
